mods/mc_tutorial/init.lua

```diff
@@ -0,0 +1,1708 @@
+        local yaw, pitch = vect_to_yp(event.dir)
+        return nil, "look in direction "..(event.dir and "(yaw = "..yaw.."°, pitch = "..pitch.."°)" or "[?]")
+    end,
+    [mc_tutorial.ACTION.POS_ABS] = function(event)
+        return nil, "go to position "..(event.pos and "(x = "..event.pos.x..", y = "..event.pos.y..", z = "..event.pos.z..")" or "[?]")
+    end,
+    [mc_tutorial.ACTION.GROUP] = function(event)
+        if event.g_type == mc_tutorial.GROUP.START then
+            return "#CCFFFF", "GROUP "..(event.g_id or "[?]").." {"
+        else
+            return "#CCFFFF", "} END GROUP "..(event.g_id or "[?]")
+        end
+    end,
+}
+
+function mc_tutorial.show_record_fs(player)
+    local pname = player:get_player_name()
+	if mc_tutorial.check_privs(player, mc_tutorial.recorder_priv_table) then
+        -- Tutorial formspec for recording a tutorial
+        local context = get_context(pname)
+        mc_tutorial.record.temp[pname] = mc_tutorial.record.temp[pname] or mc_tutorial.get_temp_shell()
+        local temp = mc_tutorial.record.temp[pname]
+
+        -- Get all recorded events
+        if not context.events then
+            local events = {}
+
+            for i,event in ipairs(temp.sequence) do
+                if event.action then
+                    local col, event_string = event_action_map[event.action](event)
+                    table.insert(events, (col or "")..minetest.formspec_escape(event_string or ""))
+                else
+                    table.insert(events, "#FFCCCC"..minetest.formspec_escape("[?]"))
+                end
+            end
+            context.events = events
+        end
+
+        -- Get all available rewards
+        if not context.rewards then
+            local rewards = {}
+            local selected_rewards = {}
+            for priv,_ in pairs(minetest.registered_privileges) do
+                if mc_helpers.tableHas(temp.on_completion.privs, priv) then
+                    table.insert(selected_rewards, {col = "#FFCCFF", s = minetest.formspec_escape("[P] ")..priv})
+                else
+                    table.insert(rewards, {col = "#FFCCFF", s = minetest.formspec_escape("[P] ")..priv})
+                end
+            end
+
+            local item_map = {
+                ["tool"] = {col = "#CCFFFF", s_pre = minetest.formspec_escape("[T] ")},
+                ["node"] = {col = "#CCFFCC", s_pre = minetest.formspec_escape("[N] ")},
+            }
+            for item,def in pairs(minetest.registered_items) do
+                local item_trim = mc_helpers.trim(item)
+                if item_trim ~= "" then
+                    local raw_col_field = item_map[def.type] or {col = "#FFFFCC", s_pre = minetest.formspec_escape("[I] ")}
+                    if mc_helpers.tableHas(temp.on_completion.items, item) then
+                        table.insert(selected_rewards, {col = raw_col_field.col, s = raw_col_field.s_pre..item_trim})
```

Comment:
Edit: we should never allow `noclip` to be granted in mc.

Reply:
> 

See my previous comment for why I disagree with never allowing `noclip `to be granted (within reason).

Reply:
> It would be helpful to check and then indicate in the formspec if there are any special privileges needed to use the tool, so that the recording player is aware that they might also need to grant that privilege when the player completes the tutorial.

This may be easier said than done, since there isn't a set way to get the privileges needed to use a tool. However, I could add the privileges `mc_toolhandler` requires for the tool to be given, along with a warning that the tool is being managed by `mc_toolhandler`

Reply:
Hm, you are right. This might be a nice request to the core developers, something like `minetest.get_tool_privs`. For now, we can at least manage the teacher controller.

Reply:
Have added warning messages to the descriptions of dangerous privileges + an additional confirmation popup for especially dangerous ones (`teacher`, `ban`, `privs`, `password`, `server`, etc.)

Will work on restricting grantable privs to those the recorder has. I think having the `privs` privilege should override this, since `privs` allows the user to grant any privileges (and likewise with `basic_privs`, although for a smaller subset of privileges)  
Would it be worthwhile to also restrict privilege reward removal in the same way? (i.e. if a previously recorded tutorial grants `priv`, editors must also have `priv` in order to remove it from the reward list)